### PR TITLE
[TensorIterator reduction improvement for mixed-precision sum on GPU]

### PR DESCRIPTION
**Pull request by jjsjann123**
Removes cast of half to float in torch.sum, with float16 input tensor and
float32 output tensor, instead we cast data when loading input in kernel.

This supposingly would save a kernel launch as well as a full global memory load
on promoted data type (float).

### CODE REVIEW

**Comment by jjsjann123** [comment]
@ngimel @csarofeen

**Comment by soumith** [comment]
cc: @colesbury

File: aten/src/ATen/native/TensorIterator.h [hunk 1]
```diff
@@ -206,6 +206,9 @@ struct CAFFE2_API TensorIterator {
   /// reductions.
   bool should_accumulate() const { return accumulate_; }
 
+  /// If the kernel should promote the output type.
+  bool should_promote_output() const { return promote_output_types_; }
```
Comment (colesbury) [inline aten/src/ATen/native/TensorIterator.h]: Is this used?

File: aten/src/ATen/native/TensorIterator.cpp [hunk 2]
```diff
@@ -475,6 +475,7 @@ std::unique_ptr<TensorIterator> TensorIterator::reduce_op(Tensor& out, const Ten
   auto builder = TensorIterator::Builder();
   builder.add_output(out);
   builder.add_input(a);
+  builder.dont_compute_common_dtype();
```
Comment (colesbury) [inline aten/src/ATen/native/TensorIterator.cpp]: I'm worried that this will fail on CPU and fail on mixed-types that aren't fp16/fp32.

Perhaps this should only be called if out is fp32 and a is fp16 and both are CUDA.

**Review by colesbury — COMMENTED** [review]
This generally looks reasonable. If I understand correctly, this code path will only be used for something like:

```python
out = torch.empty(10, dtype=torch.float32, device='cuda')
x = torch.randn(10, 10, dtype=torch.float16, device='cuda')
torch.sum(x, dim=0, out=out)
```

It  doesn't affect the following:

```python
x = torch.randn(10, 10, dtype=torch.float16, device='cuda')
torch.sum(x, dim=0)
```

Can you please add a test to test/test_cuda.py for this use case?

File: aten/src/ATen/native/TensorIterator.cpp (hunk 2 above)
Comment (jjsjann123) [reply aten/src/ATen/native/TensorIterator.cpp]: I don't think this is necessary for unary ops (hence reduction).

As I understand, reduce_op takes 1 input and 1 output, while it requires the output type to be the dtype for the operation. compute_common_dtype_ marks the promoted type operand types (`operands_[i].type`) which the op desired.
But since we are always converting the input to the desired output type (except for fp16/fp32 on gpu) https://github.com/pytorch/pytorch/pull/14580/files#diff-847680695fb0535da86db8b98a13b5a3R107, this check seems to be unnecessary.

But I can definitely add a check here if the lack of it makes us nervous.

File: aten/src/ATen/native/TensorIterator.h (hunk 1 above)
Comment (jjsjann123) [reply aten/src/ATen/native/TensorIterator.h]: Nah it's not.
Will remove it.

**Comment by jjsjann123** [comment]
Yes this PR does not affect the code path where input/output has identical dtype (or output dtype is implicit)
I thought `torch.sum(x, dim=0)` is the same thing as `x.sum(dim=0)` https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L1602-L1603

test_cuda.py already covers this part: https://github.com/pytorch/pytorch/blob/master/test/test_cuda.py#L948-L968

File: aten/src/ATen/native/TensorIterator.h [hunk 3]
```diff
@@ -228,6 +228,7 @@ struct CAFFE2_API TensorIterator {
   bool is_reduction_ = false;
   bool compute_common_dtype_ = true;
   bool allow_cpu_scalars_ = false;
+  bool promote_output_types_ = false;
```
Comment (colesbury) [inline aten/src/ATen/native/TensorIterator.h]: This also looks unused

File: aten/src/ATen/native/TensorIterator.h [hunk 4]
```diff
@@ -248,6 +249,10 @@ struct TensorIterator::Builder {
     iter_->compute_common_dtype_ = false;
   }
 
+  void promote_output_types() {
```
Comment (colesbury) [inline aten/src/ATen/native/TensorIterator.h]: And this

**Review by colesbury — COMMENTED** [review]
I don't see any tests in test_cuda.py for half -> float. At a minimum, I'd like:

1) CUDA: x.sum(dtype=float32) where x is CUDA float16 
2) CPU: x.sum(dtype=float64) where x is CPU float32
3) x.sum(out=y) where x is CPU and y is CUDA. This can raise an exception, but it must not crash

File: aten/src/ATen/native/TensorIterator.cpp (hunk 2 above)
Comment (colesbury) [reply aten/src/ATen/native/TensorIterator.cpp]: OK, I think you're right. But make sure there are unit tests for these cases.

**Comment by colesbury** [comment]
I'm not sure which tests in test_cuda.py you were linking to. The test_cuda.py link is to the `master` branch which might have changed between when you made the link and when I looked at it. In GitHub, you can press `y` when viewing a file to get a permalink to that specific version.

**Comment by jjsjann123** [comment]
Good catch. I overlooked user specified output tensors and type checking for that.

I added a sum test with fp16 input to fp32 output. Shouldn't we also put something in the test like
```import torch
x = torch.ones(65504, dtype=torch.float32).cuda()
y = torch.randn(1)
try:
  torch.sum(x, dim=[0], dtype=torch.float32, out=y)
except RuntimeError as e:
  print(e)
```
So we would know the changes crashed when it is supposed to error out.

**Comment by colesbury** [comment]
@jjsjann123 you can write that test with something like:

```python
x = torch.ones(65504, dtype=torch.float32).cuda()
y = torch.randn(1)
with self.assertRaisesRegex(RuntimeError, 'expected type torch.FloatTensor'):
  torch.sum(x, dim=[0], dtype=torch.float32, out=y)
```

https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex

File: aten/src/ATen/native/TensorIterator.cpp (hunk 2 above)
Comment (jjsjann123) [reply aten/src/ATen/native/TensorIterator.cpp]: Test cases have been added.
Also put compute_common_dtype logic back in and added another flag for type promotion from cuda.HalfTensor to cuda.FloatTensor

**Comment by jjsjann123** [comment]
Not sure what is going on with that failed CI... @colesbury any hint?

File: test/test_cuda.py [hunk 5]
```diff
@@ -1521,6 +1521,31 @@ def test_caching_pinned_memory_multi_gpu(self):
         self.assertEqual(gpu_tensor1[0], 1)
         self.assertEqual(gpu_tensor0[0], 2)
 
+    @skipIfRocm
+    def test_sum_cpu_gpu_mismatch(self):
+        x = torch.randn(20, dtype=torch.float32, device='cuda')
+        y = torch.randn(1, dtype=torch.float32)
+        try:
```
Comment (colesbury) [inline test/test_cuda.py]: Don't use this pattern.

If the torch.sum() should succeed, just use self.assertEqual. If it should fail, use one of the self.assertRaises variants.

Comment (jjsjann123) [reply test/test_cuda.py]: I can update that.
Right now the test should fail. If we ever added support for cpu_gpu mismatched input/output tensors we'll need to update this test case then.

**Comment by jjsjann123** [comment]
Earlier test failure seems to be false alarm. CI is green again @colesbury

**Review by colesbury — APPROVED** [review]
Thanks!
